File: src/main/java/net/unaussprechlich/managedgui/lib/templates/defaults/container/DefChatMessageContainer.java
/*
 * ***************************************************************************
 *
 *         Copyright © 2016 unaussprechlich - ALL RIGHTS RESERVED
 *
 * ***************************************************************************
 */

package net.unaussprechlich.managedgui.lib.templates.defaults.container;

import com.mojang.realmsclient.gui.ChatFormatting;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.minecraftforge.client.event.GuiOpenEvent;
import net.unaussprechlich.managedgui.lib.container.Container;
import net.unaussprechlich.managedgui.lib.container.callback.ICallbackUpdateHeight;
import net.unaussprechlich.managedgui.lib.databases.Player.PlayerModel;
import net.unaussprechlich.managedgui.lib.event.EnumDefaultEvents;
import net.unaussprechlich.managedgui.lib.event.util.EnumTime;
import net.unaussprechlich.managedgui.lib.event.util.Event;
import net.unaussprechlich.managedgui.lib.handler.MouseHandler;
import net.unaussprechlich.managedgui.lib.helper.DateHelper;
import net.unaussprechlich.managedgui.lib.util.EnumEventState;
import net.unaussprechlich.managedgui.lib.util.storage.ContainerSide;

import java.util.Date;

/**
 * DefChatMessageContainer Created by Alexander on 27.02.2017.
 * Description:
 **/
public class DefChatMessageContainer extends Container {

    private static int SPACE = 4;

    private DefTextListAutoLineBreakContainer message_con;
    private DefTextContainer username_con;
    private DefPictureContainer avatar_con;

    private ICallbackUpdateHeight callback;

    private final PlayerModel player;
    private final DateHelper date;

    public DefChatMessageContainer(PlayerModel player, String message, DefPictureContainer avatar_con, int width) {
        this.date = new DateHelper();
        this.player = player;
        this.avatar_con = avatar_con;
        this.username_con = new DefTextContainer(player.getRankName() + ChatFormatting.GRAY + ChatFormatting.ITALIC + "  " + date.getDateTimeTextPassed());
        setWidth(width);
        setup(message);
    }

    public void setHeightCallback(ICallbackUpdateHeight callback) {
        this.callback = callback;
    }

    public String getPlayername(){
        return player.getName();
    }

    private void setup(String message){
        avatar_con.setMargin(SPACE);
        avatar_con.setWidth(18);
        avatar_con.setHeight(18);
        username_con.setMargin(new ContainerSide().BOTTOM(4).TOP(4));
        username_con.setXOffset(avatar_con.getWidthMargin());
        message_con = new DefTextListAutoLineBreakContainer(message, getWidth() - avatar_con.getWidthMargin() - SPACE - 14, data -> update());
        message_con.setXYOffset(avatar_con.getWidthMargin(), username_con.getHeightMargin() + 2);

        registerChild(message_con);
        registerChild(avatar_con);
        registerChild(username_con);

        update();
    }

    public void addMessage(String s){
        message_con.addEntry(s);
        update();
    }

    private void update(){
        super.setHeight(SPACE * 2  + username_con.getHeightMargin() + message_con.getHeightMargin());
        message_con.setWidth(getWidth() - avatar_con.getWidthMargin() - SPACE - 14);
        if(callback != null){
            callback.call(getHeight());
        }
    }

    private static String getStyledTime(){
        Date date = new Date();
        return ChatFormatting.DARK_GRAY +"" + ChatFormatting.ITALIC + " " + date.getHours() + ":"  + date.getMinutes();
    }

    @Override
    public void setHeight(int height) {
        throw new UnsupportedOperationException("[ManagedGuiLib][DefMessageContainer] setHeight() is handled automatically use setPadding() instead!");
    }

    @Override
    protected boolean doClientTickLocal() {
        return true;
    }

    @Override
    protected boolean doRenderTickLocal(int xStart, int yStart, int width, int height, EnumEventState ees) {
        return true;
    }

    @Override
    protected boolean doChatMessageLocal(ClientChatReceivedEvent e) {
        return true;
    }

    @Override
    protected boolean doClickLocal(MouseHandler.ClickType clickType, boolean isThisContainer) {
        return true;
    }

    @Override
    protected boolean doScrollLocal(int i, boolean isThisContainer) {
        return true;
    }

    @Override
    protected boolean doMouseMoveLocal(int mX, int mY) {
        return true;
    }

    @Override
    protected <T extends Event> boolean doEventBusLocal(T iEvent) {
        if(iEvent.getID() != EnumDefaultEvents.TIME.get()) return true;
        if(iEvent.getData().equals(EnumTime.SEC_5)){
                username_con.setText(player.getRankName() + ChatFormatting.GRAY + ChatFormatting.ITALIC + "  " + date.getDateTimeTextPassed());
        }
        return true;
    }

    @Override
    protected boolean doOpenGUILocal(GuiOpenEvent e) {
        return true;
    }

    @Override
    protected boolean doResizeLocal(int width, int height) {
        update();
        return true;
    }
}

File: src/main/java/net/unaussprechlich/managedgui/lib/templates/defaults/container/DefChatMessageContainer.kt
/*
 * ***************************************************************************
 *
 *         Copyright © 2016 unaussprechlich - ALL RIGHTS RESERVED
 *
 * ***************************************************************************
 */

package net.unaussprechlich.managedgui.lib.templates.defaults.container

import com.mojang.realmsclient.gui.ChatFormatting
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.client.event.GuiOpenEvent
import net.unaussprechlich.managedgui.lib.container.Container
import net.unaussprechlich.managedgui.lib.container.callback.ICallbackUpdateHeight
import net.unaussprechlich.managedgui.lib.databases.player.PlayerModel
import net.unaussprechlich.managedgui.lib.event.EnumDefaultEvents
import net.unaussprechlich.managedgui.lib.event.util.EnumTime
import net.unaussprechlich.managedgui.lib.event.util.Event
import net.unaussprechlich.managedgui.lib.handler.MouseHandler
import net.unaussprechlich.managedgui.lib.helper.DateHelper
import net.unaussprechlich.managedgui.lib.util.EnumEventState
import net.unaussprechlich.managedgui.lib.util.storage.ContainerSide
import java.util.*

/**
 * DefChatMessageContainer Created by Alexander on 27.02.2017.
 * Description:
 */
class DefChatMessageContainer(private val player: PlayerModel, message: String, private val avatar_con: DefPictureContainer, width: Int) : Container() {

    private var message_con: DefTextListAutoLineBreakContainer? = null
    private val username_con: DefTextContainer

    private var callback: ICallbackUpdateHeight? = null
    private val date: DateHelper = DateHelper()

    init {
        this.username_con = DefTextContainer(player.rankName + ChatFormatting.GRAY + ChatFormatting.ITALIC + "  " + date.dateTimeTextPassed)
        setWidth(width)
        setup(message)
        player.getPlayerHeadLoc {avatar_con.resourceLocation = it}
    }

    fun setHeightCallback(callback: ICallbackUpdateHeight) {
        this.callback = callback
    }

    val playername: String
        get() = player.name

    private fun setup(message: String) {
        avatar_con.setMargin(SPACE)
        avatar_con.width = 18
        avatar_con.height = 18
        username_con.margin = ContainerSide().BOTTOM(4).TOP(4)
        username_con.xOffset = avatar_con.widthMargin
        message_con = DefTextListAutoLineBreakContainer(message, width - avatar_con.widthMargin - SPACE - 14) { data -> update() }
        message_con!!.setXYOffset(avatar_con.widthMargin, username_con.heightMargin + 2)

        registerChild<DefTextListAutoLineBreakContainer>(message_con)
        registerChild(avatar_con)
        registerChild(username_con)

        update()
    }

    fun addMessage(s: String) {
        message_con!!.addEntry(s)
        update()
    }

    private fun update() {
        super.setHeight(SPACE * 2 + username_con.heightMargin + message_con!!.heightMargin)
        message_con!!.width = width - avatar_con.widthMargin - SPACE - 14
        if (callback != null) {
            callback!!.call(height)
        }
    }

    override fun setHeight(height: Int) {
        throw UnsupportedOperationException("[ManagedGuiLib][DefMessageContainer] setHeight() is handled automatically use setPadding() instead!")
    }

    override fun doClientTickLocal(): Boolean {
        return true
    }

    override fun doRenderTickLocal(xStart: Int, yStart: Int, width: Int, height: Int, ees: EnumEventState): Boolean {
        return true
    }

    override fun doChatMessageLocal(e: ClientChatReceivedEvent): Boolean {
        return true
    }

    override fun doClickLocal(clickType: MouseHandler.ClickType, isThisContainer: Boolean): Boolean {
        return true
    }

    override fun doScrollLocal(i: Int, isThisContainer: Boolean): Boolean {
        return true
    }

    override fun doMouseMoveLocal(mX: Int, mY: Int): Boolean {
        return true
    }

    override fun <T : Event<*>> doEventBusLocal(iEvent: T): Boolean {
        if (iEvent.id != EnumDefaultEvents.TIME.get()) return true
        if (iEvent.data == EnumTime.SEC_5) {
            username_con.text = player.rankName + ChatFormatting.GRAY + ChatFormatting.ITALIC + "  " + date.dateTimeTextPassed
        }
        return true
    }

    override fun doOpenGUILocal(e: GuiOpenEvent): Boolean {
        return true
    }

    override fun doResizeLocal(width: Int, height: Int): Boolean {
        update()
        return true
    }

    companion object {

        private val SPACE = 4

        private val styledTime: String
            get() {
                val date = Date()
                return ChatFormatting.DARK_GRAY.toString() + "" + ChatFormatting.ITALIC + " " + date.hours + ":" + date.minutes
            }
    }
}
